src/basic/chase-symlinks.c

```diff
@@ -113,69 +114,74 @@ int chase_symlinks(
-                r = open(path, O_PATH|O_CLOEXEC|((flags & CHASE_NOFOLLOW) ? O_NOFOLLOW : 0));
+                r = openat(dir_fd, path, O_PATH|O_CLOEXEC|((flags & CHASE_NOFOLLOW) ? O_NOFOLLOW : 0));
                 if (r < 0)
                         return -errno;
 
                 *ret_fd = r;
                 return 0;
         }
 
-        if (original_root) {
-                r = path_make_absolute_cwd(original_root, &root);
+        /* When AT_FDCWD is used with a relative path, the path is taken relative to the current working
+         * directory but resolved relative to the host's root directory. */
+
+        if (dir_fd == AT_FDCWD && !path_is_absolute(path)) {
+                r = path_make_absolute_cwd(path, &buffer);
                 if (r < 0)
                         return r;
+        } else {
+                buffer = strdup(path);
+                if (!buffer)
+                        return -ENOMEM;
+        }
 
-                /* Simplify the root directory, so that it has no duplicate slashes and nothing at the
-                 * end. While we won't resolve the root path we still simplify it. Note that dropping the
-                 * trailing slash should not change behaviour, since when opening it we specify O_DIRECTORY
-                 * anyway. Moreover at the end of this function after processing everything we'll always turn
-                 * the empty string back to "/". */
-                delete_trailing_chars(root, "/");
-                path_simplify(root);
+        /* If we get a valid directory fd and an absolute path, we need to return a relative path because
+         * openat() like functions generally ignore the directory fd if provided with an absolute path. On
+         * the other hand, if we get an invalid directory fd (which includes AT_FDCWD in this case) and an
+         * absolute path, we need to return an absolute path because if we don't, if the caller is using
+         * AT_FDCWD and passes the returned relative path to another openat() like function, it would be
+         * resolved relative to the current working directory instead of to "/" like it's supposed to be. And
+         * if the caller is using -1 as the directory fd, passing the returned relative path to another
+         * openat() like function would result in an EINVAL error.
+         */
 
-                if (flags & CHASE_PREFIX_ROOT) {
-                        buffer = path_join(root, path);
-                        if (!buffer)
-                                return -ENOMEM;
-                }
+        if (dir_fd < 0 && path_is_absolute(path)) {
+                done = strdup("/");
+                if (!done)
+                        return -ENOMEM;
         }
 
-        if (!buffer) {
-                r = path_make_absolute_cwd(path, &buffer);
-                if (r < 0)
-                        return r;
-        }
+        rootfd = fd_reopen(MAX(dir_fd, -1), O_CLOEXEC|O_DIRECTORY|O_PATH);
```

Comment:
hmmm? `MAX(dir_fd, -1)`?

Reply:
To ignore `AT_FDCWD` basically, but I now see that this is rather unreadable and brittle, since it depends on `AT_FDCWD` being smaller than -1, so let me rework this